providers/asrockrack/helpers.go

```diff
@@ -80,30 +80,53 @@ func (a *ASRockRack) setFlashMode() error {
 	return nil
 }
 
+func multipartSize(fieldname, filename string) int64 {
+	body := &bytes.Buffer{}
+	form := multipart.NewWriter(body)
+	form.CreateFormFile(fieldname, filename)
+	form.Close()
+	return int64(body.Len())
+}
+
 // 2 Upload the firmware file
-func (a *ASRockRack) uploadFirmware(endpoint string, fwReader io.Reader) error {
+func (a *ASRockRack) uploadFirmware(endpoint string, fwReader io.Reader, fileSize int64) error {
 
-	// setup a buffer for our multipart form
-	var form bytes.Buffer
-	w := multipart.NewWriter(&form)
+	fieldName, fileName := "fwimage", "image"
+	contentLength := multipartSize(fieldName, fileName) + fileSize
 
-	// create form data from update image
-	fwWriter, err := w.CreateFormFile("fwimage", "image")
-	if err != nil {
-		return err
-	}
+	// setup pipe
+	pipeReader, pipeWriter := io.Pipe()
+	defer pipeReader.Close()
 
-	// copy file contents into form payload
-	_, err = io.Copy(fwWriter, fwReader)
-	if err != nil {
-		return err
-	}
+	// initiate a mulitpart writer
+	form := multipart.NewWriter(pipeWriter)
 
-	// multi-part content type
-	headers := map[string]string{"Content-Type": w.FormDataContentType()}
+	errCh := make(chan error, 1)
+	go func() {
+		defer pipeWriter.Close()
+
+		// create form part
+		part, err := form.CreateFormFile(fieldName, fileName)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		// copy from source into form part writer
+		_, err = io.Copy(part, fwReader)
+		if err != nil {
+			errCh <- err
+			return
+		}
 
-	// close multipart writer - adds the teminating boundary.
-	w.Close()
+		// add terminating boundary to multipart form
+		errCh <- form.Close()
+	}()
+
+	// multi-part content type
+	headers := map[string]string{
+		"Content-Type": form.FormDataContentType(),
+	}
 
 	// POST payload
 	_, statusCode, err := a.queryHTTPS(endpoint, "POST", pipeReader, headers, contentLength)
```

Comment:
(Bear with me as I ask possibly kind of dumb questions here...)

If `queryHTTPS()` here takes an `io.Reader`, what are we achieving by spinning up a goroutine to shuffle data from `fwReader` through a pipe instead of just passing `fwReader` directly as the argument here?

Reply:
I'm also a bit confused by the APIs involved; based on a brief test I just did with a copy of the `multipartSize()` code, it looks like the `contentLength` argument to `queryHTTPS()` is I think just a "metadata" length of some sort?  (e.g. dependent on the length of the filename itself, not the actual quantity of data in the file?)  But the actual quantity of data transferred in the POST request will ultimately have to be a function of the file size, right?

Reply:
> (Bear with me as I ask possibly kind of dumb questions here...)
> 
> If `queryHTTPS()` here takes an `io.Reader`, what are we achieving by spinning up a goroutine to shuffle data from `fwReader` through a pipe instead of just passing `fwReader` directly as the argument here?

The bytes from the io.Reader (update bin file contents) need to be wrapped within a [multipart form](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST) this is done by the goroutine which [sets up the form related fields](https://github.com/bmc-toolbox/bmclib/blob/968ea38dc70cdb76f608b1379b532c20978e949a/providers/asrockrack/helpers.go#L109) and writes to one end of the pipe passed into `queryHTTPS()`

Reply:
> I'm also a bit confused by the APIs involved; based on a brief test I just did with a copy of the multipartSize() code, it looks like the contentLength argument to queryHTTPS() is I think just a "metadata" length of some sort? (e.g. dependent on the length of the filename itself, not the actual quantity of data in the file?) But the actual quantity of data transferred in the POST request will ultimately have to be a function of the file size, right?

The [multipartSize](https://github.com/bmc-toolbox/bmclib/blob/968ea38dc70cdb76f608b1379b532c20978e949a/providers/asrockrack/helpers.go#L83) method helps figure the `Content-Length` - which includes the `firmware file size` + `multipart form fields` that have to be sent as part of the Request header - before the entire firmware file is read from disk. The filesize cannot be inferred from the io.Reader - not until all the bytes are read, I suppose?